ls1intum/Artemis
# [Feature] - Exercise statistics

### Checklist

### Motivation and Context


We want to have specific exercise related statistics

### Description

Added doughnut charts in the exercise detail page and also in a new exercise statistics page, together with other metrics interesting for exercises.
We display the score distribution, the submissions, active users, active tutors, created results, created feedback

### Steps for Testing


1. Log in to Artemis
2. Navigate into the detail page of an exercise without an releaseDate or with an releaseDate in the future
3. Verify that no doughnut charts are shown and everything looks as usual
4. Move into a exercise which has a already passed releaseDate
5. Make sure the no initial error is thrown
6. Verify that the values in the doughnut charts are correct
6.1 Average score -> go to course statistics and compare with the bar in the average score bar chart
6.2 participations -> click on the title or the doughnut chart and compare the number of participations
6.3 questions -> click on the title or the doughnut chart and count the number of questions the exercise has and how many of them have an answer which is approved 
7. Click on the Statistics button on top of the detail page
8. make sure the values of the doughnut charts are the same as before
9. compare the score distribution graph with the actual scores achieved by the students (on the scores page, accessable by clicking onto the average score doughnut chart title or graph)
10. participate in the exercise and make sure the corresponding graphs graphs behave accordingly (submit something -> bar in the submission chart increases accordingly)
11. Change the tabs (to DAY, MONTH, QUARTER or YEAR) and make sure values and labels are still correct (The doughnut charts and the score distribution graph do not change as those do not depend on a time constraint)

### Test Coverage



StatisticsResource.java: 90%
StatisticsService.java: 94%
StatisticsRepository.java: 86%

doughnut-chart.component.ts: 89%
exercise-statistics.component.ts: 100%
exercise-detail-statistics.component.ts: 100%
(Improved:
text-exercise-detail.component.ts: 98%
modeling-exercise-detail.component.ts: 89%
file-upload-exercise-detail.component.ts: 93%
programming-exercise-detail.component.ts: 57%
)

### Screenshots



The new Exercise Statistics:
<img width="1439" alt="Screenshot 2021-05-23 at 09 59 58" src="https://user-images.githubusercontent.com/44401560/119253293-cc3b2b00-bbb0-11eb-920b-95fa21e666c4.png">

3 new doughnut charts in the exercise detail view:
<img width="1439" alt="Screenshot 2021-05-23 at 10 41 14" src="https://user-images.githubusercontent.com/44401560/119253891-86339680-bbb3-11eb-8ff6-59cfe3242edb.png">
